### ISSUE TITLE
[DialogResult] Add TryAgain and Continue (result 10, and 11) respectively.

### ISSUE THREAD

**Issue by AraHaan** [dotnet/winforms#4712]
**Is your feature request related to a problem? Please describe.**

Nope, I am filing this because I noticed that the documentations here: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-messagebox
As such I noticed that DialogResults of 10, and 11 are possible.

**Describe the solution you'd like and alternatives you've considered**

Currently I have no other alternatives in mind other than adding the 2 members to ``DialogResult``.
Also might need to add ``MB_CANCELTRYCONTINUE`` to the button options too for ``Cancel``, ``Try Again``, and ``Continue`` buttons (I can file a separate issue for this one).
It also seems to miss on the default buttons to also be able to make button 4 the default if they wanted to as well (I can file a separate issue for this one as well). Nowever I would logically only make it available in .NET 6+ if approved.

**Will this feature affect UI controls?**

I do not think it will affect them at all, it will instead allow one to handle if one pressed the ``continue`` button on a dialog or ``try again`` to do specific things for those specific results which could be seen as an enhancement.

API Changes:
```diff
namespace System.Windows.Forms
{
    public enum DialogResult
    {
-        OK = 1,
+        OK = (int)ID.OK,
        // docs stay the same.
-        Cancel = 2,
+        Cancel = (int)ID.CANCEL,
        // docs stay the same.
-        Abort = 3,
+        Abort = (int)ID.ABORT,
        // docs stay the same.
-        Retry = 4,
+        Retry = (int)ID.RETRY,
        // docs stay the same.
-        Ignore = 5,
+        Ignore = (int)ID.IGNORE,
        // docs stay the same.
-        Yes = 6,
+        Yes = (int)ID.YES,
        // docs stay the same.
-        No = 7,
+        No = (int)ID.NO,
+
+        /// <summary>
+        /// The dialog box return value is Try Again (usually sent from a button labeled Try Again).
+        /// </summary>
+        TryAgain = (int)ID.TRYAGAIN,
+
+        /// <summary>
+        /// The dialog box return value is Continue (usually sent from a button labeled Continue).
+        /// </summary>
+        Continue = (int)ID.CONTINUE,
    }
    public enum MessageBoxButtons
    {
+        /// <summary>
+        ///  Specifies that the message box contains Cancel, Try Again, and Continue buttons.
+        ///  This field is constant.
+        /// </summary>
+        CancelTryContinue = (int)MB.CANCELTRYCONTINUE,
    }
    public enum MessageBoxDefaultButton
    {
+        /// <summary>
+        ///  Specifies that the forth button on the message box should be the
+        ///  default button.
+        /// </summary>
+        Button4 = (int)MB.DEFBUTTON4,
    }
}
```

As for MessageBox class this code here: https://github.com/dotnet/winforms/blob/f337d18dfcf05f9764a90895358928d6c342c4f2/src/System.Windows.Forms/src/System/Windows/Forms/MessageBox.cs#L413-L427

can be replaced with:

```diff
      try
      {
          ID mbresult = MessageBoxW(new HandleRef(owner, handle), text, caption, style);
+         return mbresult switch
+         {
+             _ => (DialogResult)mbresult,
+         };
      }
      finally
      {
          Application.EndModalMessageLoop();
          ThemingScope.Deactivate(userCookie);
          // Right after the dialog box is closed, Windows sends WM_SETFOCUS back to the previously active control
          // but since we have disabled this thread main window the message is lost. So we have to send it again after
          // we enable the main window.
          User32.SendMessageW(new HandleRef(owner, handle), User32.WM.SETFOCUS);
      }
```

And this could be attempted on removal:
https://github.com/dotnet/winforms/blob/f337d18dfcf05f9764a90895358928d6c342c4f2/src/System.Windows.Forms/src/System/Windows/Forms/MessageBox.cs#L35-L56

Edit: Shown complete changes to the enums, The changes to the ``MessageBox`` class that would be needed for this is a WIP.

**Comment by RussKie** [comment]
Thank you.

Looking at WinUser.h these appear to have been added in Windows 2000. Besides these `DialogResult` misses `IDCLOSE`, `IDHELP` and `IDTIMEOUT`:

```h
/*
 * Dialog Box Command IDs
 */
#define IDOK                1
#define IDCANCEL            2
#define IDABORT             3
#define IDRETRY             4
#define IDIGNORE            5
#define IDYES               6
#define IDNO                7
#if(WINVER >= 0x0400)
#define IDCLOSE         8
#define IDHELP          9
#endif /* WINVER >= 0x0400 */

#if(WINVER >= 0x0500)
#define IDTRYAGAIN      10
#define IDCONTINUE      11
#endif /* WINVER >= 0x0500 */

#if(WINVER >= 0x0501)
#ifndef IDTIMEOUT
#define IDTIMEOUT 32000
#endif
#endif /* WINVER >= 0x0501 */
```

We do have imports for these however in:
https://github.com/dotnet/winforms/blob/03d1550f1d893fa964548df8a6f57b2fde892fc3/src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.ID.cs#L9-L23

I think a better way to resolve this would be map `DialogResult` values to `ID` values (like in `MessageBoxButtons`), and deprecate this:
https://github.com/dotnet/winforms/blob/03d1550f1d893fa964548df8a6f57b2fde892fc3/src/System.Windows.Forms/src/System/Windows/Forms/MessageBox.cs#L35-L56

> might need to add `MB_CANCELTRYCONTINUE` to the button options too

Likewise we have a definition for this, so it'll need a mapping in `MessageBoxButtons`:
https://github.com/dotnet/winforms/blob/03d1550f1d893fa964548df8a6f57b2fde892fc3/src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.MB.cs#L12-L20

> ...also might need to add `MB_CANCELTRYCONTINUE` to the button options too...
> It also seems to miss on the default buttons to also be able to make button 4 the default if they wanted to as well...

Please update your proposal to show the complete change to the API surface.


Are you willing to champion the change and get it implemented and tested, once the API change is approved?

**Comment by AraHaan** [comment]
Alright made changes, I hope you guys do not mind if I use ``var`` or even switch expressions with or patterns to simplify it's readability.

As for the rest I think it should be complete, let me know about ``ID.CLOSE => buttons is MessageBoxButtons.CancelTryContinue or MessageBoxButtons.OKCancel or MessageBoxButtons.RetryCancel or MessageBoxButtons.YesNoCancel ? DialogResult.Cancel : DialogResult.No,`` I thought it was a good idea to make it return ``DialogResult.Cancel`` whenever the buttons are told to contain a cancel button, ``DialogResult.No`` otherwise.

I guess that part could need documentation change so the user knows it as well too.

**Comment by RussKie** [comment]
> ```diff
> +             ID.CLOSE => buttons is MessageBoxButtons.CancelTryContinue 
> +                                 or MessageBoxButtons.OKCancel 
> +                                 or MessageBoxButtons.RetryCancel 
> +                                 or MessageBoxButtons.YesNoCancel
> +                      ? DialogResult.Cancel : DialogResult.No,
> +             ID.HELP or ID.TIMEOUT => DialogResult.No,
> ```

Can you please elaborate on this piece? Under what circumstances do you expect this to happen, as according to [the docs](https://docs.microsoft.com/windows/win32/api/winuser/nf-winuser-messagebox#return-value), these don't belong to the list of the return values?

**Comment by AraHaan** [comment]
> 
> 
> > ```diff
> > +             ID.CLOSE => buttons is MessageBoxButtons.CancelTryContinue 
> > +                                 or MessageBoxButtons.OKCancel 
> > +                                 or MessageBoxButtons.RetryCancel 
> > +                                 or MessageBoxButtons.YesNoCancel
> > +                      ? DialogResult.Cancel : DialogResult.No,
> > +             ID.HELP or ID.TIMEOUT => DialogResult.No,
> > ```
> 
> Can you please elaborate on this piece? Under what circumstances do you expect this to happen, as according to [the docs](https://docs.microsoft.com/windows/win32/api/winuser/nf-winuser-messagebox#return-value), these don't belong to the list of the return values?

I think that the close button at the top of a MessageBox makes it return ``CLOSE``, if Help is enabled for it, it could possibly return ``HELP`` (it did not exactly say that it did, but it also did not say that it would not return ``HELP`` as the result of the MessageBox call) and dispatch an ``WM_HELP`` message as well too. As for ``Timeout`` the docs was not clear if ``Timeout`` was used by the MessageBox api's.

**Comment by RussKie** [comment]
Thank you. I'll get the ball rolling to get an approval. In mean time if you can - open a draft PR with the change, it would aid the process.

**Comment by AraHaan** [comment]
@RussKie ``DialogResult.None`` does not seem to be used by the MessageBox class at all, is anything using that member of ``DialogResult`` or is it just a dummy member that goes unused entirely just for backwards compatibility with old code that used it?

**Comment by AraHaan** [comment]
opened #4746.

**Comment by AraHaan** [comment]
It does not seem there are any tests for MessageBox at all except for a call I see in the TaskDialog test.

**Comment by terrajobst** [comment]
* Looks good as proposed


```C#
namespace System.Windows.Forms
{
    public partial enum DialogResult
    {
        // ...
        TryAgain,
        Continue
    }
    public partial enum MessageBoxButtons
    {
        // ...
        CancelTryContinue
    }
    public partial enum MessageBoxDefaultButton
    {
        // ...
        Button4
    }
}
```

**Comment by AraHaan** [comment]
@terrajobst so do I convert the pr from draft now for merging or does it still have some steps left that needs done before merge?

**Comment by RussKie** [comment]
The API proposal has been approved, and you now have a green light to make your PR non-draft :)

**Comment by AraHaan** [comment]
ok, converted it to a normal PR.

**Comment by AraHaan** [comment]
@terrajobst after some testing it seems setting the default button to Button4 crashes the application when trying to display a messagebox according to what @RussKie tested.

Not sure if it's just an issue with Windows however and that the crash might need to be fixed or not, it might have been an accidentally exposed field sdk headers but was never implemented.

As such the proposal might need to be edited to remove Button4, unless there is a reason why it somehow crashes applications that set it.

**Comment by terrajobst** [comment]
That's odd. Seems hard to believe that we found a bug in the SDK though. @RussKie, do we have a repro?

**Comment by RussKie** [comment]
@terrajobst no, false alarm. More details here https://github.com/dotnet/winforms/pull/4746#issuecomment-819997698

**Comment by AraHaan** [comment]
Ah ok, I thought it was a windows bug causing it to crash, good to know what the 4th button was for.

**Comment by Zheng-Li01** [comment]
@AraHaan @RussKie, verified the issue with .NET 6.0 6.0.100-preview.6.21322.9, the new function has been implemented for now, just a minor issue need to confirm that no similar **Summary** for new added buttons. see below screenshot.
![image](https://user-images.githubusercontent.com/38325459/123027570-a0f66680-d410-11eb-87c3-de45ff820bd0.png)

**Comment by AraHaan** [comment]
How is that possible?

**Comment by RussKie** [comment]
This means our intellisence bundles haven't been updated yet.

**Comment by AraHaan** [comment]
> This means our intellisence bundles haven't been updated yet.

wait there is a separate bundle?

**Comment by RussKie** [comment]
Yes, xml-docs flow through a number of other systems and teams before becoming intellisence and docs. :)
